spec/models/course_spec.rb

```diff
@@ -2150,4 +2150,164 @@
       )
     end
   end
+
+  describe "Update notification emails" do
+    let(:mailer_spy) { spy }
+
+    let(:provider) { create(:provider) }
+    let(:accrediting_provider) { create(:provider, :accredited_body) }
+    let(:organisation) { create(:organisation, providers: [provider]) }
+    let(:site_status) { create(:site_status, :findable) }
+    let(:user_one) { create(:user, organisations: [organisation]) }
+    let(:user_two) { create(:user, organisations: [organisation]) }
+
+    let(:course) do
+      create(
+        :course,
+        provider: provider,
+        accrediting_provider_code: accrediting_provider.provider_code,
+        age_range_in_years: "11_to_15",
+        qualification: "pgce_with_qts",
+        study_mode: "full_time",
+        site_statuses: [site_status],
+        maths: :equivalence_test,
+        english: :equivalence_test,
+      )
+    end
+
+    before do
+      stub_const("CourseUpdateEmailMailer", mailer_spy)
+    end
+
+    context "A self-accredited course" do
+      let(:course) { create(:course, :self_accredited, provider: accrediting_provider) }
+
+      before do
+        UserNotification.new(user: user_one, provider_code: accrediting_provider.provider_code, course_update: true).save!
+        UserNotification.new(user: user_two, provider_code: accrediting_provider.provider_code, course_update: false).save!
+      end
+
+      it "Does not send a notification" do
+        course.update!(age_range_in_years: "10_to_14")
+
+        expect(mailer_spy).not_to have_received(:course_update_email)
+      end
+    end
+
+    context "A non self-accredited course" do
+      context "With no users with notifications enabled" do
+        it "does nothing" do
+          course.update!(age_range_in_years: "10_to_14")
+
+          expect(mailer_spy).not_to have_received(:course_update_email)
+        end
+      end
+
+      context "With a user with notifications enabled" do
+        before do
+          UserNotification.new(user: user_one, provider_code: accrediting_provider.provider_code, course_update: true).save!
+          UserNotification.new(user: user_two, provider_code: accrediting_provider.provider_code, course_update: false).save!
+        end
+
+        shared_examples "Sending update notifications" do
+          it "Sends the notification to the correct user" do
+            course.assign_attributes(course_update)
+            course.ensure_site_statuses_match_study_mode
+            course.save!
+
+            expect(mailer_spy).to have_received(:course_update_email) do |course, attribute_changed, user|
+              expect(course).to eq(course)
+              expect(attribute_changed).to eq(expected_attribute_change)
+              expect(user).to eq(user_one)
+            end
```

Comment:
_But_ ... we don't test that the email was delivered, i.e. that `deliver_now` is called. Is it worth either testing a message chain or `ActionMailer::Base.deliveries`?

Reply:
:thinking: Yeah I could test that `delivery_now` is called